pkg/controller/direct/redis/cluster/rediscluster_controller.go

```diff
@@ -0,0 +1,284 @@
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/monitoring"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+)
+
+func init() {
+	registry.RegisterModel(krm.RedisClusterGVK, newRedisClusterModel)
+}
+
+func newRedisClusterModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &redisClusterModel{config: config}, nil
+}
+
+type redisClusterModel struct {
+	config *config.ControllerConfig
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &redisClusterModel{}
+
+type redisClusterAdapter struct {
+	projectID  string
+	location   string
+	resourceID string
+
+	desired *pb.Cluster
+	actual  *pb.Cluster
+
+	clustersClient *api.CloudRedisClusterClient
+}
+
+// adapter implements the Adapter interface.
+var _ directbase.Adapter = &redisClusterAdapter{}
+
+// AdapterForObject implements the Model interface.
+func (m *redisClusterModel) AdapterForObject(ctx context.Context, kube client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcpClient, err := newGCPClient(ctx, m.config)
+	if err != nil {
+		return nil, err
+	}
+	redisClustersClient, err := gcpClient.newClusterClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &krm.RedisCluster{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	location := direct.ValueOf(obj.Spec.Location)
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	projectRef, err := refs.ResolveProject(ctx, kube, obj, &obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+
+	// TODO: Move from monitoring package into shared package (and make refs implement an interface)
+	if err := monitoring.VisitFields(obj, &refNormalizer{ctx: ctx, src: obj, project: *projectRef, kube: kube}); err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desiredProto := RedisClusterSpec_ToProto(mapCtx, &obj.Spec)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	return &redisClusterAdapter{
+		projectID:      projectID,
+		location:       location,
+		resourceID:     resourceID,
+		desired:        desiredProto,
```

Comment:
Oh, and we can also group the only-KCC "desired" field to structs for different purposes to normalize the solution, like the externalRef (real ID) https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/cloudbuild/externalresource.go#L28

Reply:
Right!